tools/generalize.cpp

```diff
@@ -0,0 +1,163 @@
+#include "llvm/Support/MemoryBuffer.h"
+#include "llvm/Support/GraphWriter.h"
+#include "llvm/Support/KnownBits.h"
+
+#include "souper/Infer/Preconditions.h"
+
+#include "souper/Inst/InstGraph.h"
+#include "souper/Parser/Parser.h"
+#include "souper/Tool/GetSolver.h"
+#include "souper/Util/DfaUtils.h"
+
+using namespace llvm;
+using namespace souper;
+
+unsigned DebugLevel;
+
+static cl::opt<unsigned, /*ExternalStorage=*/true>
+DebugFlagParser("souper-debug-level",
+     cl::desc("Control the verbose level of debug output (default=1). "
+     "The larger the number is, the more fine-grained debug "
+     "information will be printed."),
+     cl::location(DebugLevel), cl::init(1));
+
+static cl::opt<std::string>
+InputFilename(cl::Positional, cl::desc("<input souper optimization>"),
+              cl::init("-"));
+
+static llvm::cl::opt<bool> RemoveLeaf("remove-leaf",
+    llvm::cl::desc("Try to generalize a valid optimization by replacing"
+                   "the use of a once-used variable with a new variable"
+                   "(default=false)"),
+    llvm::cl::init(false));
+
+static llvm::cl::opt<bool> FixIt("fixit",
+    llvm::cl::desc("Given an invalid optimization, generate a valid one."
+                   "(default=false)"),
+    llvm::cl::init(false));
+
+void Generalize(InstContext &IC, Solver *S, ParsedReplacement Input) {
+  bool FoundWP = false;
+  std::vector<std::map<Inst *, llvm::KnownBits>> Results;
+  S->abstractPrecondition(Input.BPCs, Input.PCs, Input.Mapping, IC, FoundWP, Results);
+
+  if (FoundWP && Results.empty()) {
+    Input.print(llvm::outs(), true);
+  } else {
+    for (auto &&Result : Results) { // Each result is a disjunction
+      for (auto Pair: Result) {
+        Pair.first->KnownOnes = Pair.second.One;
+        Pair.first->KnownZeros = Pair.second.Zero;
+      }
+      Input.print(llvm::outs(), true);
+    }
+  }
+}
+
+// TODO: Return modified instructions instead of just printing out
+void RemoveLeafAndGeneralize(InstContext &IC,
+                               Solver *S, ParsedReplacement Input) {
+
+  if (DebugLevel > 1) {
+  llvm::errs() << "Attempting to generalize by removing leaf.\n";
+  }
+  // TODO: Do not generalize by removing leaf if LHS has one inst.
+
+  std::map<Inst *, std::set<Inst *>> Uses;
+
+  std::vector<Inst *> Stack{Input.Mapping.LHS, Input.Mapping.RHS};
+  // TODO: Find uses in PCs/BPCs
+
+  std::set<Inst *> Visited;
+  while (!Stack.empty()) {
+   auto Current = Stack.back();
```

Comment:
indentation error.

if you're having trouble with basic code formatting, why not get into the habit of running clang-format on new files before contributing them?

Reply:
Good idea, I'll do them cmake tweaking required for running `ninja format` or `make format`.

Reply:
for now, please only do this on new files that are being added to the repo. doing this for existing files would be fun, but it would make it effectively impossible to ever merge any existing branch. we can only do this at a time where we're certain that no non-trivial branches exist. at least one exists right now, and probably more.